proposals/partial-properties.md

```diff
@@ -0,0 +1,142 @@
+
+```cs
+partial class C
+{
+    // Defining declaration
+    public partial string Prop { get; set; }
+
+    // Implementing declaration
+    public partial string Prop { get => field; set => field = value; }
+}
+```
+
+Because we have reserved the syntactic form with semicolon accessor bodies for the *defining declaration*, a partial property cannot be *automatically implemented*. We therefore adjust [Automatically implemented properties (§15.7.4)](https://github.com/dotnet/csharpstandard/blob/standard-v7/standard/classes.md#1574-automatically-implemented-properties) as follows:
+
+> An automatically implemented property (or auto-property for short), is a non-abstract, non-extern, **non-partial,** non-ref-valued property with semicolon-only accessor bodies.
+
+**Remarks**. It is useful for the compiler to be able to look at a single declaration in isolation and know whether it is a defining or an implementing declaration. Therefore we don't want to permit auto-properties by including two identical `partial` property declarations, for example. We don't think that the use cases for this feature involve implementing the partial property with an auto-property, but in cases where a trivial implementation is desired, we think the `field` keyword makes things simple enough.
+
+A partial property must have one *defining declaration* and one *implementing declaration*.
+
+**Remarks**. We also don't think it is useful to allow splitting the declaration across more than two parts, to allow different accessors to be implemented in different places, for example. Therefore we simply imitate the scheme established by partial methods.
+
+Similar to partial methods, the attributes in the resulting property are the combined attributes of the parts are concatenated in an unspecified order, and duplicates are not removed.
+
+Only the defining declaration of a partial property participates in lookup, similar to how only the defining declaration of a partial method participates in overload resolution.
+
+**Remarks**. In the compiler, we would expect that only the symbol for the defining declaration appears in the member list, and the symbol for the implementing part can be accessed through the defining symbol. However, some features like nullable analysis might *see through* to the implementing declaration in order to provide more useful behavior.
+
+```cs
+partial class C
+{
+    public partial string Prop { get; set; }
+    public partial string Prop { get => field; set => field = value; }
+
+    public C() // warning CS8618: Non-nullable property 'Prop' must contain a non-null value when exiting constructor. Consider declaring the property as nullable.
+    {
+    }
+}
+```
+
+### Matching signatures
+The property declarations must have the same type and ref kind. The property declarations must have the same set of accessors.
```

Comment:
"must have the same type"

I think this should be adjusted. For example, the following should be allowed:
```
public partial string { get => ""; }
public partial string? { get; }
```

We should follow what partial methods do.

Reply:
My preference here would be to use similar requirements as the "partial method 'strict' comparison", except to report any difference under this comparison as an error. Essentially, let's tolerate nullable oblivious differences, but almost nothing else, as much as is practical.

I'll try and look for a straightforward way to specify this.

Reply:
Keep in mind that one of our goals of NRT is that user code and generated code can adopt NRT independently of each other. A `partial` member is squarely in the generated code scenario. The decision we make here needs to facilitate users and generators adopting NRT at different times. 

Given that I think it makes the most sense to allow NRT differences between the `partial` definition and declaration.

Reply:
I think it would be reasonable to:
1. give no diagnostics on nullable-oblivious difference. This addresses the needs of source generators.
2. give a *warning* diagnostic on any nullable-enabled difference (annotated vs. not-annotated).
3. The declaration part "wins" re: what types are actually used in the signature of the property for use sites and emit, just as with partial method declarations.

The reason I phrased (2) in this particular way is, I don't want to have to specify the conditions under which a nullability difference is considered "safe" across partial property declarations, which would include considerations around the set of accessors (is the property readable? writable?) and so on.

Reply:
If you don't want to know about nrt, then you need to disable nullable, and you will never get a warning resulting from (2).

Reply:
Consider the case where a generator doesn't know about NRT. Now the developer adds NRT annotations to a partial property. If the warnings are in the generated file then only the generator can fix them. That goes against the "adopt independently" goal.

Reply:
I would stick to the same rules as partial methods (which I think are the rules of overrides, see `SourceMemberContainerTypeSymbol.CheckValidNullableMethodOverride` used in `PartialMethodChecks`):
```
    // no warning
    public partial string? M();
    public partial string M() { return ""; }

    // warning
    public partial string M();
    public partial string? M() { return ""; }
```
With the exception of null-oblivious, no variance would be allowed for a property with both getter and setter, but some variance would be allowed for read-only or write-only properties.

Reply:
```cs
    // no warning
    public partial string? M();
    public partial string M() { return ""; }
```
Try this in a console app. It is a warning.

Reply:
What I am intending to propose at this point, is something which falls entirely along the precedent that we have for partial methods, in terms of which scenarios receive warnings or errors by default, and where those diagnostics are reported. This is a precedent which we have shipped and through the following years have found that it works for all the usage scenarios that we care about. There is nothing about the nature of partial properties which fundamentally changes the question of which signature differences to warn/error about or not, or where to report the warnings.

The only difference I am putting forward on the properties side is that I want to simplify the design in comparison to partial methods, by not having both "strict" versus "legacy" mode of comparison, and essentially only having one mode equivalent to "strict" for methods. This is because the strict comparison for partial methods was only added later and in a warning wave for compat purposes. There is no such compat concern for partial properties.

Reply:
I'd tried in sharplab (no warning) but in VS I see the warning (CS8826 `WRN_PartialMethodTypeDifference`). Sharplab seems to not have warning waves turned on...
Sounds good. Thanks for the clarification :-)